### DISCUSSION TITLE
Need guidance regarding GitHub LFS

### DISCUSSION THREAD

**Discussion by GitAccountNcc** [community/community#141665] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

General

### Body

My requirement is to pull a repository and have a back up of it in my local system at scheduled intervals. I need to handle Larger Files in GitHub repository as well.
I am able to pull the new LFS files. But the LFS file which are updated or deleted is not reflecting my local system.
I need to handle nearly 1000 repositories in each schedule. So I cannot determine the extension of the larger files, depth of the repository(folder level) or handle the larger files one by one.
It will be great if i get any guidance on the same.

**Comment by Mnahil-Naseer** [comment]


**Reply by GitAccountNcc** [reply]
Even after running git lfs fetch and git lfs checkout, i am not able to find the LFS files deletion or update. Kindly let me know if there is any other approach to handle the larger files.

**Reply by GitAccountNcc** [reply]
The following set of commands works fine for pull and deletion of larger files. 
"git stash save && git lfs ls-files && git pull && git lfs pull && git config -l | grep lfs && git lfs checkout && git lfs fetch --all && git lfs prune". The only issue i am facing now is, the larger files are getting pulled every time, even if it is unchanged.  
For PULL scenario, my requirement is to fetch the newly added files, updated files and also to delete the file in local that is already deleted in the repository. But now even if there is no change, all the larger files are fetched in each pull, which is increasing the usage of bandwidth. Kindly let me know how to fix this.

**Comment by milansamuel609** [comment]
# 1) Use a Script for Automation
You can write a script (using languages like Python or Bash) that automates the following:

- Clone or pull updates from the repositories.
- Handle Git LFS files (fetch updates, including deleted or modified files).
- Schedule the script to run periodically (using a scheduler like cron on Linux or Task Scheduler on Windows).

# 2. Handling Git LFS Files
Git LFS requires special attention because it manages large files differently. You need to ensure that your backup script also pulls the latest LFS files. Here's a way to handle this:


Clone a repository with LFS:

```bash
git clone --recurse-submodules <repo_url>
git lfs fetch --all
```

Update existing repository with LFS:

```bash
git pull
git lfs pull
```

Ensure all LFS changes are captured:

```bash
git fetch --all
git lfs fetch --all
git lfs pull
```

# 3) Backup Script Example (Using Python)

Below is a Python example using GitPython to manage multiple repositories. It also handles LFS updates: 

```bash
import os
import subprocess
from git import Repo
from datetime import datetime

# Paths to store backup
BASE_BACKUP_PATH = "/path/to/your/backup/directory"
REPOSITORIES = [
    "https://github.com/user/repo1.git",
    "https://github.com/user/repo2.git",
    # Add more repositories as needed
]

# Function to clone or update a repo
def backup_repo(repo_url):
    repo_name = repo_url.split("/")[-1].replace(".git", "")
    repo_path = os.path.join(BASE_BACKUP_PATH, repo_name)

    try:
        if not os.path.exists(repo_path):
            # Clone the repository with LFS support
            print(f"Cloning {repo_name}...")
            subprocess.run(["git", "clone", "--recurse-submodules", repo_url, repo_path])
        else:
            # Pull the latest updates
            print(f"Updating {repo_name}...")
            repo = Repo(repo_path)
            repo.git.pull()
            
            # Fetch LFS updates
            subprocess.run(["git", "fetch", "--all"], cwd=repo_path)
            subprocess.run(["git", "lfs", "fetch", "--all"], cwd=repo_path)
            subprocess.run(["git", "lfs", "pull"], cwd=repo_path)
        
        print(f"Backup complete for {repo_name}.")
    except Exception as e:
        print(f"Error backing up {repo_name}: {e}")

# Backup all repositories
for repo_url in REPOSITORIES:
    backup_repo(repo_url)
```

# 4) Scheduling the Backup
For scheduling, here are some options:

- Linux (cron): Schedule the Python script using a cron job. Example to run every day at 2 AM:

```bash
0 2 * * * /usr/bin/python3 /path/to/your/script.py
```

Windows (Task Scheduler): Create a task that triggers the Python script at your desired interval.

# 5) Git LFS Considerations

- Ensure Git LFS is installed on the system running the script: git lfs install.
- Storage: Make sure you have enough local storage to handle LFS files, as they can be quite large.
- Connection Stability: Since you’re dealing with 1000 repositories, network reliability is crucial for pulling large LFS files.

# 6) Error Handling & Logging
Given the scale of your task:

- Implement logging in the script to capture errors.
- Add retries in case of network failures.
- Use .gitignore to ignore unnecessary files during backup if possible.

# 7) Alternative Tools
Consider tools that handle large-scale Git management:

- GitHub CLI (gh): You can use gh for managing multiple repositories.
- BFG Repo-Cleaner: If you need to clean up LFS files.

# 8) Use .gitattributes for LFS
Ensure .gitattributes is correctly set up in your repositories to track file types for LFS:

``` bash

*.zip filter=lfs diff=lfs merge=lfs -text
*.jpg filter=lfs diff=lfs merge=lfs -text
# Add more file types as needed
```

This approach should allow you to automate the backup of multiple repositories while properly handling large files tracked by Git LFS.

Please kindly try all the above steps given. This may help you to fix your issue.
